### PR TITLE
new backend: uptobox

### PR DESCRIPTION
**Pull request by buengese**
Yet another backend from me this time it's Uptobox as requested in #4467. I'm almost done with this backend barring some minor things and docs.

I am however somewhat hesitant to merge this backend. Uptobox is a storage provider with rather questionable reputation being strongly associated with piracy resulting in it being delisted from Google. That hasn't sopped me from using them (it's very cheap "unlimited" storage after all) but it's something one should be aware of.  

Other than that it comes with some quirks common for these kinds of file-sharing services most annoyingly it's not possible to upload to any other folders than the root folder. Also, no modtime or checksum though there would be a way to hack in modtime using the description field.
@ncw 

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: backend/uptobox/uptobox.go [hunk 1]
```diff
@@ -0,0 +1,950 @@
...
+//
+// dir should be "" to list the root, and should not have
+// trailing slashes.
+//
+// This should return ErrDirNotFound if the directory isn't
+// found.
+func (f *Fs) List(ctx context.Context, dir string) (entries fs.DirEntries, err error) {
+	root := f.dirPath(dir)
+
+	var limit uint64 = 10 // max number of objects per request
```
Comment (ncw) [inline backend/uptobox/uptobox.go:317]: Seems very low?

File: backend/uptobox/uptobox.go [hunk 2]
```diff
@@ -0,0 +1,950 @@
...
+		return errors.Wrap(err, "couldn't update file info")
+	}
+	if info.StatusCode != 0 {
+		return errors.Errorf("updateFileInfo: api error: %d - %s", info.StatusCode, info.Message)
+	}
+	return err
+}
+
+func (f *Fs) putUnchecked(ctx context.Context, in io.Reader, remote string, size int64, options ...fs.OpenOption) (fs.Object, error) {
+	if size > int64(200e9) { // max size 200G
```
Comment (ncw) [inline backend/uptobox/uptobox.go:478]: GiB or GB or unknown?

File: backend/uptobox/uptobox.go [hunk 3]
```diff
@@ -0,0 +1,950 @@
...
+	upload, err := f.uploadFile(ctx, in, size, tmpName, info.Data.UploadLink, options...)
+	if err != nil {
+		return nil, err
+	}
+	if len(upload.Files) != 1 {
+		return nil, errors.New("Upload: unexpected response")
+	}
+	match := f.IDRegexp.FindStringSubmatch(upload.Files[0].URL)
+
+	// move file to destionation folder
```
Comment (ncw) [inline backend/uptobox/uptobox.go:512]: spelling!

File: backend/uptobox/uptobox.go [hunk 4]
```diff
@@ -0,0 +1,950 @@
...
+		return errors.Errorf("remove: api error: %d - %s", info.StatusCode, info.Message)
+	}
+	return nil
+}
+
+// Check the interfaces are satisfied
+var (
+	_ fs.Fs     = (*Fs)(nil)
+	_ fs.Copier = (*Fs)(nil)
+	_ fs.Mover  = (*Fs)(nil)
```
Comment (ncw) [inline backend/uptobox/uptobox.go:948]: No DirMove?

**Review by ncw — COMMENTED** [review]
I gave the code a brief run through - see inline comments

> Yet another backend from me this time it's Uptobox as requested in #4467. I'm almost done with thisbackend barring some minor things and docs.

:-)

> I am however somewhat hesitent to merge this backend. Uptobox is storage provider with rather questionable reputation being strongly associated with piracy resulting in it being delisted from google. That hasn't sopped me from using them (it's very cheap "unlimited" storage afterall) but it's something one should be aware of.

Hmm, yes there are some [interesting reddit threads](https://www.reddit.com/r/webhosting/comments/aej5yy/uptoboxcom_will_remove_your_life_for_free/)

> Also in their terms of service it says this:
> 
> > Doing massive cold storage (your files are mainly never downloaded) is not allowed. Hot storage is unlimited.

So it really for sharing files, not storing them.

I think as long as we point that out to the user in the backend docs then I don't see why we shouldn't have a backend for them. I don't want to make disappointed rclone users who back up there and have to make their backups public, but maybe people want to upload their Linux ISO collection there?

> Other than that it comes with some quirks common for services most annoyingly it's not possible to upload to any other folders than the root folder.

I noted your 7 API calls comment!

> Also no modtime or checksum though there would be a way to hack in modtime using the description field.

Backends without modtime are annoying, though you can use `--update` on them usually.

We've steered clear of adding metadata to descriptions I think. It isn't a terrible idea though - I'd quite like to be able to store more general purpose metadata with things (think a dictionary of stuff) for which I'd imagined we'd take over description fields. Maybe leave it for the moment?

**Comment by buengese** [comment]
- list object limit increased to 100 (maximum)
- file size limit 200GB
- hopefully fixed spelling
- DirMove added

File: backend/uptobox/uptobox.go [hunk 5]
```diff
@@ -0,0 +1,1052 @@
...
+
+func init() {
+	fs.Register(&fs.RegInfo{
+		Name:        "uptobox",
+		Description: "Uptobox",
+		Config: func(ctx context.Context, name string, config configmap.Mapper) {
+		},
+		NewFs: NewFs,
+		Options: []fs.Option{{
+			Help: "Your API Key, get it from https://uptobox.com/my_account",
```
Comment (ncw) [inline backend/uptobox/uptobox.go:48]:
```
Your API Key, get it from https://uptobox.com/my_account
Enter a string value. Press Enter for the default ("").
```

The URL printed is correct , but the thing you need to get is called the `token`

image

File: docs/content/uptobox.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+title: "Uptobox"
+description: "Rclone docs for Uptobox"
+---
+
+{{< icon "fa fa-archive" >}} Uptobox
+-----------------------------------------
+
+This is a Backend for Uptobox file storage service. Uptobox is closer to a one-click hoster than a traditional 
+cloud storage provider and therefore not suitable for long term storage. Please note that a Premium account is
+required to use this backend.
```
Comment (ncw) [inline docs/content/uptobox.md:11]: This seemed to work with a free account for me?

**Review by ncw — APPROVED** [review]
I think this is looking good to me.

The backend passed my smoke tests!

I put a couple of comments inline which may/may not need addressing!

Feel free to merge when you are happy :-)

File: docs/content/uptobox.md (hunk 6 above)
Comment (buengese) [reply docs/content/uptobox.md:11]: That's rather surprising to me. The docs state that you need to request a waiting token and wait up to 30 minutes between downloads. I'm completely skipping this step which supposedly only works when you have a premium account. Maybe that doesn't apply to files you uploaded yourself? I'll have to test that.

**Pull request merged**